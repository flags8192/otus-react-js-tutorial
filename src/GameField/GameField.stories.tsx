import React, { FC, useState, useEffect } from 'react';
import { withKnobs, number, boolean } from '@storybook/addon-knobs';
import { GameField } from '@src/GameField/index';
import { CellInfo } from '@src/Cell/models';
import { getInverted } from '@src/Cell/utils';

export default {
  component: GameField,
  decorators: [withKnobs],
  title: 'Example/GameField',
};

const START_WIDTH = 3;
const START_HEIGHT = 3;
const START_DATA: CellInfo[] = [];

export const Dynamic: FC = () => {
  const width = number('Width, cells', START_WIDTH);
  const height = number('Height, cells', START_HEIGHT);
  const showAll = boolean('showAll', false);

  const [data, setData] = useState(START_DATA);

  useEffect(() => {
    const cellsNumber = width * height;
    const startCellState = showAll ? CellInfo.alive : CellInfo.dead;
    const newData: CellInfo[] = [];
    for (let i = 0; i < cellsNumber; i++) {
      newData.push(startCellState);
    }
    setData(newData);
  }, [width, height, showAll]);

  const handleCellClick = (num: number) => {
    if (num < 0 || num >= data.length) {
      return;
    }

    const newData = data.concat();
    newData[num] = getInverted(newData[num]);
    setData(newData);
  };

  return <GameField data={data} onCellClick={handleCellClick} width={width} />;
};
