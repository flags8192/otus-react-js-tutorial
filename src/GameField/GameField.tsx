import React, { FC } from 'react';
import { CellInfo } from '@src/Cell/models';
import { Cell } from '@src/Cell';
import styled from '@emotion/styled';

export const FieldContainer = styled.section`
  background: #fff;
`;

export interface IGameField {
  data: CellInfo[];
  onCellClick: (num: number) => void;
  width: number;
}

export const GameField: FC<IGameField> = ({ onCellClick, width, data }) => {
  const widthMinus1 = width - 1;
  const heightMinus1 = Math.floor(data.length / width) - 1;
  return (
    <FieldContainer role="cellsGrid">
      {data.map((item, index) => {
        const y = Math.floor(index / width);
        const x = index % width;
        const isRight = x === widthMinus1;
        const isBottom = y === heightMinus1;
        const isLeft = x === 0;
        return (
          <Cell
            key={`${y}-${x}`}
            num={index}
            alive={item === CellInfo.alive}
            onClick={onCellClick}
            isLeft={isLeft}
            isRight={isRight}
            isBottom={isBottom}
          />
        );
      })}
      <div style={{ clear: 'both' }} />
    </FieldContainer>
  );
};
