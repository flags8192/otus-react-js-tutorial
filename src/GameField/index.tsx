export * from '@src/GameField/GameField';
