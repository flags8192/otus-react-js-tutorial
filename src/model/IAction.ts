import IPayload from './IPayload';

export interface IAction {
  type: string;
  payload?: IPayload;
}
