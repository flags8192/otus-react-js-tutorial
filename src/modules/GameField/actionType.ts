const prefix = 'GAME_FIELD__';

export default {
  SET_SIZE: `${prefix}SET_SIZE`,
};
