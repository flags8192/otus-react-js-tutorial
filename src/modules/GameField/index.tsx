import React, { FC, useContext } from 'react';
import styled from '@emotion/styled';

import { ICellInfo } from '@src/modules/Cell/models';
import Cell from '@src/modules/Cell';
import store from '@src/store';

export const FieldContainer = styled.section`
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  margin: auto;
  height: fit-content;
  width: fit-content;
`;

const GameField: FC = () => {
  const { state } = useContext(store);
  const { gameFieldData, gameFieldSize, selectedCells = {} } = state;
  const widthMinus1 = gameFieldSize - 1;
  const heightMinus1 = Math.floor(gameFieldData.length / gameFieldSize) - 1;

  if (!gameFieldSize) {
    return <p role="dataMessage">Нет данных!</p>;
  }

  return (
    <FieldContainer role="cellsGrid">
      {gameFieldData.map((item: ICellInfo, index: number) => {
        const y = Math.floor(index / gameFieldSize);
        const x = index % gameFieldSize;
        const isRight = x === widthMinus1;
        const isBottom = y === heightMinus1;
        const isLeft = x === 0;
        return (
          <Cell
            key={`${y}-${x}`}
            num={index}
            isFilled={item}
            isSelected={selectedCells[index]}
            isLeft={isLeft}
            isRight={isRight}
            isBottom={isBottom}
          />
        );
      })}
      <div style={{ clear: 'both' }} />
    </FieldContainer>
  );
};

export default GameField;
