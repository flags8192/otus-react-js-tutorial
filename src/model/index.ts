import { IState } from '@src/model/IState';
import { IAction } from '@src/model/IAction';

export { IAction, IState };
