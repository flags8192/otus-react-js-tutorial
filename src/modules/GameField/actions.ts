import actionType from './actionType';

const setGameFieldSize = (gameFieldSize: number) => ({
  type: actionType.SET_SIZE,
  payload: { gameFieldSize },
});

export { setGameFieldSize };
