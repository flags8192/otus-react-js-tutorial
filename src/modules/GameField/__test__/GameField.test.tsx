import React from 'react';
import GameField from '../index';
import { render, screen } from '@testing-library/react';
import Store from '@src/store';
import { IAction, IState } from '@src/model';

let dispatch: (action: IAction) => void;
let state: IState;

describe('GameField', () => {
  beforeEach(() => {
    dispatch = jest.fn();
    state = {
      gameLevel: '1',
      gameFieldSize: 3,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
      selectedCells: {},
    };
  });

  it('renders 9 cells from data:[9 items]', () => {
    const { asFragment } = render(
      <Store.Provider value={{ dispatch, state }}>
        <GameField />
      </Store.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const cellsGrid = screen.getByRole(/cellsGrid/gi);
    expect(cellsGrid).toBeInTheDocument();
    const cellContainer = screen.getAllByRole(/cellContainer/gi);
    expect(cellContainer.length).toBe(5);
  });
});
