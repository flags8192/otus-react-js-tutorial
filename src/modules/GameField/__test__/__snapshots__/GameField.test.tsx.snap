// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameField renders 9 cells from data:[9 items] 1`] = `
<DocumentFragment>
  <section
    class="css-hsjawn"
    role="cellsGrid"
  >
    <div
      class="css-es8rv7"
      role="cellContainer-0"
    >
      <div
        class="css-w6oewy"
        role="cellContent-0"
      />
    </div>
    <div
      class="css-2r18jt"
      role="cellContainer-1"
    />
    <div
      class="css-1jbr29x"
      role="cellContainer-2"
    />
    <div
      class="css-es8rv7"
      role="cellContainer-3"
    />
    <div
      class="css-2r18jt"
      role="cellContainer-4"
    />
    <div
      class="css-1jbr29x"
      role="cellContainer-5"
    />
    <div
      class="css-es8rv7"
      role="cellContainer-6"
    />
    <div
      class="css-2r18jt"
      role="cellContainer-7"
    />
    <div
      class="css-1jbr29x"
      role="cellContainer-8"
    />
    <div
      style="clear: both;"
    />
  </section>
</DocumentFragment>
`;
